### PR TITLE
Dali TF installation: check import before completing the installation

### PR DESCRIPTION
**Pull request by jantonguirao**
#### Why we need this PR?
- It adds new feature needed to make the installation of DALI TF plugin more reliable.

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     *Added an import check to the DALI TF installation script to verify that the installed plugin is usable before completing the installation sucessfully*
 - Affected modules and functionalities:
     *DALI TF installer*
 - Key points relevant for the review:
     *NA*
 - Validation and testing:
     *Existing tests apply*
 - Documentation (including examples):
     *NA*

**JIRA TASK**: *[DALI-2170]*

### CODE REVIEW

File: dali_tf_plugin/dali_tf_plugin_install_tool.py [hunk 1]
```diff
@@ -97,14 +97,30 @@ def debug_str(self):
 
         return s
 
+    def check_import(self, lib_path):
+        import tensorflow as tf
+        import nvidia.dali as dali  # DALI symbols need to be loaded
```
Comment (JanuszL) [inline dali_tf_plugin/dali_tf_plugin_install_tool.py:102]: So we cannot install DALI and plugin in one go as DALI needs to be fully available here, right?

**Review by JanuszL — APPROVED** [review]

File: dali_tf_plugin/dali_tf_plugin_install_tool.py [hunk 2]
```diff
@@ -97,14 +106,30 @@ def debug_str(self):
 
         return s
 
+    def check_import(self, lib_path):
+        import tensorflow as tf
+        assert can_import_dali()  # DALI symbols need to be loaded
```
Comment (JanuszL) [inline dali_tf_plugin/dali_tf_plugin_install_tool.py:111]: Assert or just false?
We just want to fail if there is no exact match in this case or build from source anyway?

File: dali_tf_plugin/dali_tf_plugin_install_tool.py [hunk 3]
```diff
@@ -97,14 +106,30 @@ def debug_str(self):
...
+
     def install_prebuilt(self):
         assert(self.can_install_prebuilt)
         assert(self.prebuilt_plugin_best_match is not None)
         assert(self.plugin_name is not None)
         print("Tensorflow was built with g++ {}, providing prebuilt plugin".format(self.tf_compiler))
-        plugin_dest =  os.path.join(self.plugin_dest_dir, self.plugin_name)
-        print("Copy {} to {}".format(self.prebuilt_plugin_best_match, self.plugin_dest_dir))
-        copyfile(self.prebuilt_plugin_best_match, plugin_dest)
+        if self.check_import(self.prebuilt_plugin_best_match):
```
Comment (JanuszL) [inline dali_tf_plugin/dali_tf_plugin_install_tool.py:125]: Even if `best_version == self.tf_version` you still wants to check import, righ?

Comment (JanuszL) [reply dali_tf_plugin/dali_tf_plugin_install_tool.py:125]: In such case installing DALI together with the plugin will just fail, as there is no DALI available and you cannot `check_import` that seems to be executed in every code path now.

File: dali_tf_plugin/dali_tf_plugin_install_tool.py (hunk 2 above)
Comment (jantonguirao) [reply dali_tf_plugin/dali_tf_plugin_install_tool.py:111]: The idea is that you only call check_import if you can import the plugin. It should only be called if can_import_dali()==True. I'll correct the logic

File: dali_tf_plugin/dali_tf_plugin_install_tool.py (hunk 3 above)
Comment (jantonguirao) [reply dali_tf_plugin/dali_tf_plugin_install_tool.py:125]: That's right, check_import is executed even for exact TF version. Then we should only check import if we can. I'll correct

**Review by JanuszL — APPROVED** [review]

File: dali_tf_plugin/dali_tf_plugin_install_tool.py [hunk 4]
```diff
@@ -91,20 +97,50 @@ def debug_str(self):
...
 
         return s
 
+    def check_import(self, lib_path, dali_stub):
+        import tensorflow as tf
+        lib_name = os.path.basename(lib_path)
+        dali_stub_name = os.path.basename(dali_stub)
+        print("Importing the TF library to check for errors")
+
+        # The DALI TF lib and the DALI stub lib should be at the same directory for check_import to succeed
```
Comment (jantonguirao) [inline dali_tf_plugin/dali_tf_plugin_install_tool.py:113]: Open for a better solution

**Comment by jantonguirao** [comment]
!build

**Review by banasraf — APPROVED** [review]

File: dali_tf_plugin/dali_tf_plugin_install_tool.py [hunk 5]
```diff
@@ -38,6 +38,8 @@ def __init__(self, plugin_dest_dir = None):
         self.platform_machine = platform.machine()
         self.is_compatible_with_prebuilt_bin = self.platform_system == 'Linux' and self.platform_machine == 'x86_64'
         self.prebuilt_dir = os.path.join(self.src_path, 'prebuilt')
+        self.prebuilt_stub_dir = os.path.join(self.prebuilt_dir, 'stub')
+        self.prebuilt_dali_stub = find('libdali.so', self.prebuilt_stub_dir)[0] or None
```
Comment (JanuszL) [inline dali_tf_plugin/dali_tf_plugin_install_tool.py:42]: Is it possible to not have stub?

File: dali_tf_plugin/dali_tf_plugin_install_tool.py [hunk 6]
```diff
@@ -91,20 +97,50 @@ def debug_str(self):
...
+        dali_stub_name = os.path.basename(dali_stub)
+        print("Importing the TF library to check for errors")
+
+        # The DALI TF lib and the DALI stub lib should be at the same directory for check_import to succeed
+        # Unfortunately the symlink is necessary because we can't change LD_LIBRARY_PATH from within the script
+        with tempfile.TemporaryDirectory(prefix="check_import_tmp") as tmpdir:
+            lib_path_tmpdir = os.path.join(tmpdir, lib_name)
+            os.symlink(lib_path, lib_path_tmpdir)
+
+            dali_stub_tmpdir = os.path.join(tmpdir, dali_stub_name)
```
Comment (JanuszL) [inline dali_tf_plugin/dali_tf_plugin_install_tool.py:119]: `dali_stub_tmlink` ?

File: dali_tf_plugin/dali_tf_plugin_install_tool.py [hunk 7]
```diff
@@ -91,20 +97,50 @@ def debug_str(self):
...
+    def check_import(self, lib_path, dali_stub):
+        import tensorflow as tf
+        lib_name = os.path.basename(lib_path)
+        dali_stub_name = os.path.basename(dali_stub)
+        print("Importing the TF library to check for errors")
+
+        # The DALI TF lib and the DALI stub lib should be at the same directory for check_import to succeed
+        # Unfortunately the symlink is necessary because we can't change LD_LIBRARY_PATH from within the script
+        with tempfile.TemporaryDirectory(prefix="check_import_tmp") as tmpdir:
+            lib_path_tmpdir = os.path.join(tmpdir, lib_name)
```
Comment (JanuszL) [inline dali_tf_plugin/dali_tf_plugin_install_tool.py:116]: `lib_path_tmplink`?

**Pull request merged**